[Lens] Update field list help and empty message

dej611 opened the pull request:
## Summary

Update the `Available fields` help message when no sampling strategy is used for the list:

<img width="450" alt="Screenshot 2022-03-07 at 15 46 18" src="https://user-images.githubusercontent.com/924948/157056780-24d0b342-1176-47b0-b4fd-519e7eaa49e9.png">

In case of empty data the field popup content has been updated as follow:

<img width="580" alt="Screenshot 2022-03-03 at 17 48 26" src="https://user-images.githubusercontent.com/924948/156613939-8092b2a0-6cab-44a2-9aaf-a326028a34c2.png">

(note that in the pic the runtime field chosen is always returning no value, but has been marked as Available anyway)

If the user flips the setting to use the old sampling strategy then the old message is reused:

<img width="589" alt="Screenshot 2022-03-03 at 17 49 57" src="https://user-images.githubusercontent.com/924948/156614114-46b7edc4-b16c-4f84-a57d-f7d2cb3c75fb.png">


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

flash1293 commented:
The first sentence isn't 100% precise - in some situations (date range and constant keyword fields), the available/empty check is doing better than "entire data view". However it's still possible the field is just defined in the mapping but don't hold data. Not sure how to phrase it though. Maybe "Available fields are defined on the indices which are matching the current query and time range"?

dej611 commented:
> The first sentence isn't 100% precise - in some situations (date range and constant keyword fields), the available/empty check is doing better than "entire data view". However it's still possible the field is just defined in the mapping but doesn't hold data. Not sure how to phrase it though. Maybe "Available fields are defined on the indices which are matching the current query and time range"?

I find the message not easy to decypher to be honest.
What about:
```
Available fields in the data view. This can be influenced by current query and time range. Some field types...
```

The first part hints, in my opinion, hints to the user that the mapping is involved somehow. While the second part adds some more clues about the influence of filters on it.

dej611 commented:
Updated message 👌

dej611 commented:
@KOTungseth I have some questions for 1 and 3:

> Drag and drop available fields to the workspace and create visualizations. To change the available fields, select a different data view, edit your queries, or use a different time range. ...

In here (from 1) I see what I can do with available fields, but not necessarily what "Available fields" means.

> Available fields contain data in the first 500 documents that match your filters. ...

In here (3) I see what "Available fields" is, but no mention of how it works.

Both 1 and 3 refers to the same section, just changing depending on the way the list is fetched. While I think 3 provides enough information about the list content, I'm not sure 1 does.

KOTungseth commented:
> @KOTungseth I have some questions for 1 and 3:
> 
> > Drag and drop available fields to the workspace and create visualizations. To change the available fields, select a different data view, edit your queries, or use a different time range. ...
> 
> In here (from 1) I see what I can do with available fields, but not necessarily what "Available fields" means.
> 
> > Available fields contain data in the first 500 documents that match your filters. ...
> 
> In here (3) I see what "Available fields" is, but no mention of how it works.
> 
> Both 1 and 3 refers to the same section, just changing depending on the way the list is fetched. While I think 3 provides enough information about the list content, I'm not sure 1 does.

I updated my comment for the first UI component. Does this make more sense?

dej611 commented:
> I updated my comment for the first UI component. Does this make more sense?

@KOTungseth I think there's a misunderstanding.

Both (1) and (3) apply to the same element: the `Available fields` fields list component in Lens.
The messages will change based on a settings the user can change, which will have the effect to change *how* the list is built.

We used to have only (3) as the strategy was "fetch 500 documents, based on current filters, and workout which one has data", but recently we introduced a new strategy, hence the new message (1). The new approach looks at the configuration of the data view and then tweak it based on current filters, so it's more robust than the previous strategy.

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@KOTungseth I've updated the message based on your initial content:

UI component | Recommended changes 
-- | -- 
 <img width="262" alt="Screen Shot 2022-03-07 at 10 50 50 AM" src="https://user-images.githubusercontent.com/38990715/157080540-e50a077e-9aa5-48cb-9c83-b43ffacb9120.png"> | Drag and drop available fields to the workspace and create visualizations. To change the available fields, select a different data view, edit your queries, or use a different time range. Some field types cannot be visualized in Lens, including full text and geographic fields.
<img width="307" alt="Screen Shot 2022-03-07 at 10 50 59 AM" src="https://user-images.githubusercontent.com/38990715/157082004-e4b0efd4-6d09-4438-a08d-474998b117e5.png"> | Lens is unable to create visualizations with this field because it does not contain data. To create a visualization, drag and drop a different field.
<img width="264" alt="Screen Shot 2022-03-07 at 10 51 06 AM" src="https://user-images.githubusercontent.com/38990715/157082564-500e749c-d378-4cbd-94ca-724e430b8b96.png"> | Available fields contain the data in the first 500 documents that match your filters. To view all fields, expand Empty fields. You are unable to create visualizations with full text, geographic, flattened, and object fields.
<img width="311" alt="Screen Shot 2022-03-07 at 10 51 13 AM" src="https://user-images.githubusercontent.com/38990715/157083021-907ab21d-496c-4018-b1b1-392e20a573fc.png"> | Lens is unable to create visualizations with this field because it does not contain data in the first 500 documents that match your filters. To create a visualization, drag and drop a different field.

For (1) there's no longer the problem to explain that some fields with values can land in `Empty fields`, so the focus is now only about what user can do. (3) has been left as it was before.

dej611 commented:
@elasticmachine merge upstream

dej611 commented:
@elasticmachine merge upstream

flash1293 commented:
@elasticmachine merge upstream

flash1293 reviewed (approved):
LGTM, code review only. The messages make sense to me

flash1293 commented:
@KOTungseth I would like to merge this one, is the new wording OK?

gchaps reviewed (approved):
Yes, text LGTM.
